File: src/2020/Day 02/P2.rb
data = File.read("input.txt").split("\n")
total = 0

for line in data
    fLine = line.split(" ")

    pos = fLine[0].split("-").map{|x| Integer(x)}
    token = fLine[1][0]
    password = fLine[2].split("")

    if password[pos[0] - 1] === token && password[pos[1] - 1] === token
        next
    elsif password[pos[0] - 1] === token || password[pos[1] - 1] === token
        total += 1
    end
end

puts "Result: #{total}"

File: src/2021/Day_10/P2.js
const fs = require(`fs`);
const path = require(`path`);

const pts = {
    [`)`]: 1,
    [`]`]: 2,
    [`}`]: 3,
    [`>`]: 4
};

const pairs = {
    [`(`]: `)`,
    [`[`]: `]`,
    [`{`]: `}`,
    [`<`]: `>`
};

const openings = Object.keys(pairs);

const main = async () => {
    fs.readFile(path.resolve(__dirname, `./input.txt`), `utf-8`, (err, data) => {
        if (err) throw err;

        const lines = data.split(`\n`).map(x => x.trim());
        const totals = [];

        for (const line of lines) {
            let total = 0;

            const opened = [];
            const tokens = line.split(``);

            let isCorrupted = false;
            for (const token of tokens) {
                if (openings.includes(token)) opened.push(token);
                else if (pairs[opened.pop()] !== token) {
                    isCorrupted = true;
                    break;
                }
            }

            if (isCorrupted) continue;

            for (const token of opened.reverse()) {
                total *= 5;
                total += pts[pairs[token]];
            }

            totals.push(total);
        }

        const total = totals.sort((a, b) => a - b)[Math.floor(totals.length / 2)];
        console.log(`Result: ${total}`);
    });
};

void main();
